### ISSUE TITLE
unused-extern-crate false positive with sysroot crates

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#69304]
To use certain crates such as `proc_macro` it is required to include it with `extern crate proc_macro`. Similar thing applies to other sysroot crates.

Denying the `unused` group of lints will prevent code from compiling properly with an error like this:

```
error: `extern crate` is not idiomatic in the new edition
 --> src/lib.rs:3:1
  |
3 | extern crate proc_macro;
  | ^^^^^^^^^^^^^^^^^^^^^^^^ help: convert it to a `use`
  |
  = note: `-D unused-extern-crates` implied by `-D unused`
```

**Comment by nagisa** [comment]
Might be reproducible only when `cargo check`ing a proc_macro crate. I am having trouble in quickly minimizing this.

**Comment by nagisa** [comment]
Note that at least the "convert to `use`" suggestion is not applicable, because `use`-ing a sysroot crate does _not_ in fact work cc @estebank

**Comment by ehuss** [comment]
Hm, this is an unexpected consequence due to https://github.com/rust-lang/cargo/pull/7700.  As of beta (1.42), `extern crate proc_macro;` isn't necessary.

I can see how that might be problematic for people who use `deny(warnings)` or `deny(unused)`.  I'm not sure what the policy is for that.  For now you can either remove the `extern crate proc_macro`, or if you need compatibility with older versions, add an explicit `allow`.

**Comment by Mark-Simulacrum** [comment]
Generally speaking, we're free to break deny(warnings). Where possible, I think we usually try to phase new warnings in such that the nightly that introduces the warning has a previously released stable which supports some fix for it (e.g., deprecation issued as a "use this" would start warning once the replacement lands in stable).

**Comment by nagisa** [comment]
Ah, I see why this is happening, now.

What we do is we run nightly `cargo clippy`, which, I presume, does add the `--extern proc_macro` but we compile actual code with plain rustc and we do not add the flag (yet). Hence the disconnect between what the lint is saying and the behaviour we are seeing when compiling code.

Closing as not-a-bug.
